MaxText/maxtext_utils.py

```diff
@@ -91,30 +91,60 @@ def get_train_input_output_trees(func, input_args, input_kwargs):
   return p_train_step
 
 
-# https://arxiv.org/pdf/2204.02311.pdf Appendix B
-def calculate_tflops_training_per_device(num_model_parameters, config, log=True):
+def calculate_tflops_training_per_device(config, log=True):
   """Calculate training TFLOP"""
-  learnable_weight_tflops = 6 * num_model_parameters * config.max_target_length * config.per_device_batch_size / 10**12
-  noncasual_attention_flops = (
-      12
-      * config.num_query_heads
-      * config.num_decoder_layers
-      * config.head_dim
-      * config.max_target_length**2
+  ffn1_flops = (
+      2
       * config.per_device_batch_size
-      / 10**12
+      * config.max_target_length
+      * config.mlp_dim
+      * config.emb_dim
+      * len(config.mlp_activations)
   )
-  causal_attention_tflops = noncasual_attention_flops / 2  # due to causality in attention
-  total_tflops = learnable_weight_tflops + causal_attention_tflops
+  ffn2_flops = 2 * config.per_device_batch_size * config.max_target_length * config.mlp_dim * config.emb_dim
+  total_ffn_flops = ffn1_flops + ffn2_flops
+
+  if config.num_experts > 1:
+    # MoE: brute force implementation
+    gate_flops = 2 * config.per_device_batch_size * config.max_target_length * config.emb_dim * config.num_experts
+    weight_combination_flops = 2 * config.per_device_batch_size * config.max_target_length * config.emb_dim
```

Comment:
Based on a conversation with Ran, I believe this is accurate but not in Megatron. Because it is a small effect and needlessly drives divergence from Megatron, we will remove it from MaxText. I would love for someone at NV to confirm.

Reply:
Happy to have a look at this

Reply:
Thanks @abhinavgoel95! What's your thoughts about `weight_combination_flops`?

Reply:
I am going to remove the `weight_combination_flops` as I did not see difference from tflops. Both are 6946.5 tflops.